File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/configuration/rabbitmq/docs/serdes/ProductInfoSerDes.kt
package io.micronaut.configuration.rabbitmq.docs.serdes

// tag::imports[]
import io.micronaut.configuration.rabbitmq.bind.RabbitConsumerState
import io.micronaut.configuration.rabbitmq.intercept.MutableBasicProperties
import io.micronaut.configuration.rabbitmq.serdes.RabbitMessageSerDes
import io.micronaut.context.annotation.Requires
import io.micronaut.core.convert.ConversionService
import io.micronaut.core.type.Argument

import javax.inject.Singleton
import java.nio.charset.Charset
// end::imports[]

@Requires(property = "spec.name", value = "ProductInfoSerDesSpec")
// tag::clazz[]
@Singleton // <1>
class ProductInfoSerDes(private val conversionService: ConversionService<*>)// <3>
    : RabbitMessageSerDes<ProductInfo> { // <2>

    override fun deserialize(consumerState: RabbitConsumerState, argument: Argument<ProductInfo>): ProductInfo? { // <4>
        val body = String(consumerState.body, CHARSET)
        val parts = body.split("\\|".toRegex())
        if (parts.size == 3) {
            var size: String? = parts[0]
            if (size == "null") {
                size = null
            }

            val count = conversionService.convert(parts[1], Long::class.java)
            val sealed = conversionService.convert(parts[2], Boolean::class.java)

            if (count.isPresent && sealed.isPresent) {
                return ProductInfo(size, count.get(), sealed.get())
            }
        }
        return null
    }

    override fun serialize(data: ProductInfo?, properties: MutableBasicProperties): ByteArray { // <5>
        return (data?.size + "|" + data?.count + "|" + data?.sealed).toByteArray(CHARSET)
    }

    override fun supports(argument: Argument<ProductInfo>): Boolean { // <6>
        return argument.type.isAssignableFrom(ProductInfo::class.java)
    }

    companion object {
        private val CHARSET = Charset.forName("UTF-8")
    }
}
// end::clazz[]

File: docs-examples/example-python/src/test/python/micronaut/rabbitmq/docs/serdes/ProductInfoSerDes.py
from micronaut.context.annotation import Requires
# tag::imports[]
from jakarta.inject import Singleton
from java.lang import Boolean, Long
from micronaut.core.convert import ConversionService
from micronaut.core.type import Argument
from micronaut.rabbitmq.bind import RabbitConsumerState
from micronaut.rabbitmq.intercept import MutableBasicProperties
from micronaut.rabbitmq.serdes import RabbitMessageSerDes

from .ProductInfo import ProductInfo
# end::imports[]


@Requires(property="spec.name", value="ProductInfoSerDesSpec")
# tag::clazz[]
@Singleton  # <1>
class ProductInfoSerDes(RabbitMessageSerDes[ProductInfo]):  # <2>

    def __init__(self, conversion_service: ConversionService):  # <3>
        self.conversion_service = conversion_service

    def deserialize(self, consumer_state: RabbitConsumerState, argument: Argument[ProductInfo]) -> ProductInfo | None:  # <4>
        body = bytes(consumer_state.getBody()).decode("utf-8")
        parts = body.split("|")
        if len(parts) == 3:
            size = parts[0]
            if size == "null":
                size = None

            count = self.conversion_service.convert(parts[1], Long)
            sealed = self.conversion_service.convert(parts[2], Boolean)

            if count.isPresent() and sealed.isPresent():
                return ProductInfo(size, count.get(), sealed.get())
        return None

    def serialize(self, data: ProductInfo | None, properties: MutableBasicProperties) -> bytes | None:  # <5>
        if data is None:
            return None
        size = "null" if data.size is None else data.size
        sealed = str(data.sealed).lower()
        return f"{size}|{data.count}|{sealed}".encode("utf-8")

    def supports(self, argument: Argument[ProductInfo]) -> bool:  # <6>
        return argument.getType().isAssignableFrom(ProductInfo)
# end::clazz[]
